### DISCUSSION TITLE
Can I remove the (1) that results after using rdkit.Chem.PandasTools.WriteSDF so it works with Maestro?

### DISCUSSION THREAD

**Discussion by DKchemistry** [rdkit/rdkit#5465] [Q&A] [answered]
Hi, 

When I use `rdkit.Chem.PandasTools.WriteSDF(df, properties=list(df.columns))`, I can successfully write an SDF that Maestro can understand. 

The properties are written like this: 

```
>  <ID>  (1)
Molecule1 
```

Which is perfectly fine for most properties. 

However, for reasons I don't understand, Maestro does not like the '(1)' that follow for the property that controls subgroups. 

If I use the following code: 

`df.insert(3, 's_m_subgroup_title', pd.Series(1, index=df.index))`
`df.insert(4, 's_m_subgroupid', pd.Series(1, index=df.index))`
`rdkit.Chem.PandasTools.WriteSDF(df, out='test.sdf',properties=list(df.columns))`

I get the following ouput if I open test.sdf in a text editor 

```
>  <s_m_subgroup_title>  (1) 
1
```

```
>  <s_m_subgroupid>  (1) 
1

```
For some reason, Maestro does not like this '(1)' notation and won't add the subgroups correctly in the GUI. 

If I open a text editor and manually delete the '(1)' like such

```
>  <s_m_subgroup_title>  
1

>  <s_m_subgroupid> 
1
```

It works. Is there a way I can remove this '(1)' from ever being written to the output?

**Comment by ptosco** [comment, marked as answer]
@DKchemistry Simplest way would be to strip the number with a regex:
```
from io import StringIO
import re

strip_number = re.compile(r"^(>  <\S+>)(\s+\(\d+\)\s*)$", re.MULTILINE)

with StringIO() as buf:
    PandasTools.WriteSDF(df, buf, properties=list(df.columns))
    sdf = buf.getvalue()
sdf = strip_number.sub(r"\1", sdf)
with open("file.sdf", "w") as hnd:
    hnd.write(sdf)
```

**Reply by DKchemistry** [reply]
Thank you @ptosco! It works perfectly. I am novice to programming so understanding what's going on here was tricky, but really really helpful! While I could imagine the regex, I wouldn't have ever come up with 

```
with StringIO() as buf:
    PandasTools.WriteSDF(df, buf, properties=list(df.columns))
    sdf = buf.getvalue()
```
I'll try to keep that in mind in the future :)

**Comment by ricrogz** [comment]
Hi @DKchemistry, I'd like to solve this issue on the other end of things, in Maestro. Would you be able to share an example file or code snippet that can be used to reproduce the issue, so we can file a bug report with Maestro?

**Reply by DKchemistry** [reply]
Sure, that would be great! let me know what they say?  

Unfortunately, GitHub won't let me upload the files as .sdf so that you can copy them directly. I've changed the file ending to '.txt', I think you can just change it back to .sdf without issue. 

Phenol.txt is the output of importing a phenol into Maestro, grouping it as 'TestGroup', and then exporting it as an sdf (then me changing it to the .txt). 

If we look at the sdf file in a text editor, we see the properties look like this: 

```
> <s_m_entry_id>
1

> <s_m_entry_name>
entry

> <s_m_subgroup_title>
TestGroup

> <s_m_subgroupid>
TestGroup

> <b_m_subgroup_collapsed>
0
```

Next, I run this (reading the Maestro sdf into rdkit, then exporting it back out): 

```
import rdkit as rdkit
from rdkit.Chem import PandasTools
import pandas as pd

Original_Phenol_SDF = rdkit.Chem.PandasTools.LoadSDF(r"Phenol.sdf")

PandasTools.WriteSDF(Original_Phenol_SDF, 'rdkit_Phenol_SDF.sdf', properties=list(Original_Phenol_SDF.columns))
```
Again, can't save the sdf here directly, so saved it as .txt file. 

If I open it in a text editor like the example above, the properties field looks like this: 

```
>  <s_m_entry_id>  (1) 
1

>  <s_m_entry_name>  (1) 
entry

>  <s_m_subgroup_title>  (1) 
TestGroup

>  <s_m_subgroupid>  (1) 
TestGroup

>  <b_m_subgroup_collapsed>  (1) 
0

>  <ID>  (1) 
Phenol
```

I've opened both sdf files in maestro, you can see which Phenol belongs to which sdf by looking at the source file
![image](https://user-images.githubusercontent.com/81147591/181591016-da691078-be82-451e-8927-d602eb1c8e08.png)

Only the original SDF retains the group. 

Now, if I edit out the (1) in the rdkit sdf and load it in maestro, I can get the grouping functionality back: 

![image](https://user-images.githubusercontent.com/81147591/181591491-2c31d7e7-1045-43a1-888e-1a6cfa3cc54c.png)

Hope this helps! 

[Phenol.txt](https://github.com/rdkit/rdkit/files/9212577/Phenol.txt)
[rdkit_Phenol_SDF.txt](https://github.com/rdkit/rdkit/files/9212579/rdkit_Phenol_SDF.txt)

**Reply by ricrogz** [reply]
Thanks! I have created a ticked for the issue, and attached your files and a link to this discussion.

Sadly, the 2022-3 Release of the Schrodinger Suite is due very soon, so the fix won't make it in in time, but I'm pretty sure this will be fixed in the next one.

In the mean time, you can keep using @ptosco's code to strip the numbers from the properties.

**Reply by ricrogz** [reply]
I'm sorry that it took so long... but we've finally merged a patch that solves this issue in Maestro. It should be available in the next release of the Schrodinger Suite (2024-2).

**Reply by DKchemistry** [reply]
Thank you :)
